components/brave_wallet/browser/siwe_message_parser.cc

```diff
@@ -0,0 +1,402 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/browser/siwe_message_parser.h"
+
+#include <vector>
+
+#include "base/strings/strcat.h"
+#include "base/strings/string_tokenizer.h"
+#include "base/strings/string_util.h"
+#include "brave/components/brave_wallet/common/eth_address.h"
+#include "brave/components/brave_wallet/common/sign_message_request.mojom.h"
+#include "url/gurl.h"
+#include "url/origin.h"
+#include "url/url_constants.h"
+
+using base::StringTokenizer;
+
+namespace brave_wallet {
+
+namespace {
+constexpr size_t kMinNonceLength = 8;
+constexpr char kStartingToken[] =
+    " wants you to sign in with your Ethereum account:";
+constexpr char kURIToken[] = "URI: ";
+constexpr char kVersionToken[] = "Version: ";
+constexpr char kChainIdToken[] = "Chain ID: ";
+constexpr char kNonceToken[] = "Nonce: ";
+constexpr char kIssuedAtToken[] = "Issued At: ";
+constexpr char kExpirationTimeToken[] = "Expiration Time: ";
+constexpr char kNotBeforeToken[] = "Not Before: ";
+constexpr char kRequestIdToken[] = "Request ID: ";
+constexpr char kResourcesToken[] = "Resources:";
+constexpr char kResourcesSeperator[] = "- ";
```

Comment:
While these are all fully correct to the spec, I suspect someone will implement a validator that's more lenient than our implementation which means that the user will go through the normal 191 message flow even though the message is intended to be a SIWE message.

It may be useful to add some backup normalization catchers to detect when this is happening. Looking at the [reference implementation ](https://github.com/spruceid/siwe/blob/main/test/parsing_positive.json) though they're not so it's fine for us to opt out of this as well.

Reply:
We can display warning when fallback to 191 if we detect that is very likely to be a SIWE message but failed to conform to the spec. However we should not allow anything to slip through. If we want to do the warning, it should be follow up

Reply:
SGTM - I've opened a separate issue for this

https://github.com/brave/brave-browser/issues/32819